### PR TITLE
DBZ-6478 Blog on Debezium signaling and notifications - Customisation

### PR DESCRIPTION
**Pull request by ani-sha**
https://issues.redhat.com/browse/DBZ-6478

### CODE REVIEW

**Comment by ani-sha** [comment]
Source code for example: https://github.com/debezium/debezium-examples/pull/325

File: _posts/2023-07-10-custom-http-signaling-notification.adoc [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+featured: true
+author: anmohant
+---
+
+Welcome to this series of articles dedicated to signaling and notifications in Debezium! This post serves as the second installment in the series, where we will discuss how to customize the signal and notification channels in Debezium.
+
+Debezium 2.3 introduced new improvements in signaling and notification capabilities. You can set up new signals and notification channels in addition to the pre-defined signals and notification channels offered by Debezium. This feature enables users to customize the system to suit their unique needs and combine it with their existing infrastructure or third-party solutions.
+It enables effective monitoring and a proactive response to data changes by precisely capturing and communicating signal events and triggering notifications through preferred channels.
+
+The first article in this series, [Signaling and Notifications in Debezium](https://debezium.io/blog/2023/06/27/Debezium-signaling-and-notifications/), provides an overview of the signaling and notification features in Debezium. It also discusses the available channels & their use cases for various scenarios.
```
Comment (mfvitale) [inline _posts/2023-07-10-custom-http-signaling-notification.adoc]: On the preview, links seems not rendering well. Can you please check?

File: _posts/2023-07-10-custom-http-signaling-notification.adoc [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
++++<!-- more -->+++
+
+== Custom Signal & Notification Channel
+
+In Debezium, it is possible to customize the signal and notification channels to fit specific requirements. For example, we can achieve a customization by creating an `HTTP` channel for both signals and notifications. This `HTTP` channel receives signals from an http endpoint, and notifications can be sent back to the endpoint upon the delivery of the signal.
+
+Let's explore an example that demonstrates how to create and utilize the `HTTP` signal and notification channels using the Debezium Postgres connector, a Mock Server, and [Postbin](https://www.toptal.com/developers/postbin/) which is used as the HTTP endpoint to receive the notifications.
```
Comment (mfvitale) [inline _posts/2023-07-10-custom-http-signaling-notification.adoc]: On the preview, links seems not rendering well. Can you please check?

File: _posts/2023-07-10-custom-http-signaling-notification.adoc [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+      "body": "{\"id\":\"924e3ff8-2245-43ca-ba77-2af9af02fa07\",\"type\":\"log\",\"data\":{\"message\": \"Signal message received from http endpoint.\"}}",
+      "statusCode": 200
+    }
+  }
+]
+----
+
+The `initializerJson.json` file defines a mock http request to the path `/api/signal` with the query string parameter `code=10969`. When this request is received by the Mock Server, it will respond with a JSON body containing an `id`, `type`, and `data`. The response has a status code of 200, indicating a successful response.
+
+Create the `HTTP` signal channel by extending the `SignalChannelReader` interface as shown below:
```
Comment (mfvitale) [inline _posts/2023-07-10-custom-http-signaling-notification.adoc]:
```suggestion
Create the `HTTP` signal channel by implementing the `SignalChannelReader` interface as shown below:
```

File: _posts/2023-07-10-custom-http-signaling-notification.adoc [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+    }
+}
+----
+
+<1> The `name()` method returns the name of the signal channel. To enable Debezium to use the channel, specify the name `http` in the connector’s `signal.enabled.channels` property.
+<2> The `init()` method can be used to initialize specific configuration, variables, or connections that the http channel requires.
+<3> The `read()` method reads signal from the http endpoint and returns a list of `SignalRecord` objects which will be processed by the Debezium connectors.
+<4> The `close()` method closes all allocated resources.
+
+Create the notification channel by extending the `NotificationChannel` interface as shown below:
```
Comment (mfvitale) [inline _posts/2023-07-10-custom-http-signaling-notification.adoc]:
```suggestion
Create the notification channel by implementing the `NotificationChannel` interface as shown below:
```

File: _posts/2023-07-10-custom-http-signaling-notification.adoc [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+  {
+    "httpRequest" : {
+      "method" : "GET",
+      "path" : "/api/signal",
+      "queryStringParameters" : {
+        "code" : ["10969"]
+      }
+    },
+    "httpResponse" : {
+      "body": "{\"id\":\"924e3ff8-2245-43ca-ba77-2af9af02fa07\",\"type\":\"log\",\"data\":{\"message\": \"Signal message received from http endpoint.\"}}",
```
Comment (mfvitale) [inline _posts/2023-07-10-custom-http-signaling-notification.adoc:92]: Maybe here you can also specify what this signal is expected to do.

File: _posts/2023-07-10-custom-http-signaling-notification.adoc [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+== Custom Signal & Notification Channel
+
+In Debezium, it is possible to customize the signal and notification channels to fit specific requirements. For example, we can achieve a customization by creating an `HTTP` channel for both signals and notifications. This `HTTP` channel receives signals from an http endpoint, and notifications can be sent back to the endpoint upon the delivery of the signal.
+
+Let's explore an example that demonstrates how to create and utilize the `HTTP` signal and notification channels using the Debezium Postgres connector, a Mock Server, and [Postbin](https://www.toptal.com/developers/postbin/) which is used as the HTTP endpoint to receive the notifications.
+
+=== Setting up the `HTTP` Signal Channel:
+
+- Configure the Debezium Postgres connector to emit signals whenever a relevant database change occurs.
+- Set up a Mock Server as a receiver for these signals. The Mock Server acts as an intermediary, capturing the signals and forwarding them to the desired http endpoint.
```
Comment (mfvitale) [inline _posts/2023-07-10-custom-http-signaling-notification.adoc]: Maybe here the role of `Mock Server` is not so clear. Maybe you can clarify better that this can be any service that return `signals` but for this specific example we used `Mock Server`.

File: _posts/2023-07-10-custom-http-signaling-notification.adoc [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+- Configure the Mock Server to forward the signals to an http endpoint using the appropriate HTTP method (e.g., POST).
+- Customize the `HTTP` channel settings to define the HTTP endpoint URL, authentication, headers, and any additional parameters as needed.
+
+=== Setting up the `HTTP` Notification Channel:
+
+- Once the signal is received by the Mock Server, it can trigger the posting of a notification to an http endpoint using Postbin or any other suitable http endpoint.
```
Comment (mfvitale) [inline _posts/2023-07-10-custom-http-signaling-notification.adoc]: Also here, in a more generic way the `HttpNotificationChannel` simply sends Debezium `Notification` to an HTTP endpoint, that for this example is `Postbin`.

File: _posts/2023-07-10-custom-http-signaling-notification.adoc [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+    }
+
+    @Override
+    public void init(CommonConnectorConfig connectorConfig) { // <2>
+        // custom configuration
+    }
+
+    @Override
+    public List<SignalRecord> read() { // <3>
+        // logic to read & process signal event from http endpoint
```
Comment (mfvitale) [inline _posts/2023-07-10-custom-http-signaling-notification.adoc]: Why not putting the real implementation?

File: _posts/2023-07-10-custom-http-signaling-notification.adoc [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+
+++++
+<div class="imageblock centered-image">
+    <img src="/assets/images/http_notification_postbin.png" class="responsive-image" alt="Postbin Preview">
+</div>
+++++
+
+== Conclusion
+
+In this tutorial, we explored how to create custom signal and notification channels for Debezium connectors. We created a custom signal channel that send a signal event to the connector via a http endpoint. We also created a custom notification channel that sends a notification event to a third-party service when the signal event is processed. We used the Mock Server to simulate the third-party service and Postbin to view the notification event.
```
Comment (mfvitale) [inline _posts/2023-07-10-custom-http-signaling-notification.adoc]:
```suggestion
In this tutorial, we explored how to create custom signal and notification channels for Debezium connectors. We created a custom signal channel that receive a signal event from an HTTP endpoint. We also created a custom notification channel that sends a notification event to an HTTP endpoint.
```
WDYT to rephrase in this way?

**Review by mfvitale — CHANGES_REQUESTED** [review]
Generally LGTM. I leaved just few comments.

File: _posts/2023-07-10-custom-http-signaling-notification.adoc (hunk 8 above)
Comment (ani-sha) [reply _posts/2023-07-10-custom-http-signaling-notification.adoc]: I wasn't sure to add it here or not, as the implementation is little longer.

Comment (mfvitale) [reply _posts/2023-07-10-custom-http-signaling-notification.adoc]: Ok.

Comment (ani-sha) [reply _posts/2023-07-10-custom-http-signaling-notification.adoc]: And we are providing this example's source code inside `debezium/examples`. But would also like to know @Naros's opinion on this.

File: _posts/2023-07-10-custom-http-signaling-notification.adoc [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+----
+
+Add the following dependency to the `pom.xml` file with the Debezium version (2.3 and later versions):
+
+[source,xml]
+----
+<dependency>
+    <groupId>io.debezium</groupId>
+    <artifactId>debezium-core</artifactId>
+    <version>${debezium.version}</version>
```
Comment (Naros) [inline _posts/2023-07-10-custom-http-signaling-notification.adoc]:
```suggestion
    <version>2.3.0.Final</version>
```
It's generally better to provide bits that can be copied directly without variables for clarity.

File: _posts/2023-07-10-custom-http-signaling-notification.adoc [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+    }
+  }
+]
+----
+
+The `initializerJson.json` file defines a mock http request to the path `/api/signal` with the query string parameter `code=10969`. When this request is received by the Mock Server, it will respond with a JSON body containing an `id`, `type`, and `data`. The response has a status code of 200, indicating a successful response.
+
+The `id` is an arbitrary unique string that identifies a signal instance.
+The `type` specifies the type of signal to send. In this example, the type is `log`. It requests the connector to add an entry to the log. After processing the signal, the connector prints the specified message to the log.
+The `data` specifies JSON-formatted parameters to pass to a signal event. In this example, the `message` parameter is passed to the signal event.
```
Comment (Naros) [inline _posts/2023-07-10-custom-http-signaling-notification.adoc]:
```suggestion
`id`:: an arbitrary unique string that identifies a signal instance.
`type`:: the type of signal to send. In this example, the type is `log`, and it requests the connector to add an entry to the connector's log file. After the signal is processed, the connector prints the specified message in the log.
`data`:: the JSON-formatted parameters to pass to a signal event. In this example, the `message` parameter is passed to the signal event.
```
Description lists tend to be easier to read after a code segment than paragraphs.  If you'd prefer to use the style in the two code blocks below, that's also acceptable.  I'd suggest whichever style you pick, just be consistent, I'm fine with either.

File: _posts/2023-07-10-custom-http-signaling-notification.adoc (hunk 8 above)
Comment (Naros) [reply _posts/2023-07-10-custom-http-signaling-notification.adoc]: I concur with Mario, it's generally better from a follow-along style to include it, and given the full source isn't overly verbose I think it's fine.  Much like you've already done, you can always omit the imports to make the code more concise for the reader.

File: _posts/2023-07-10-custom-http-signaling-notification.adoc [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+    },
+    "httpResponse" : {
+      "body": "{\"id\":\"924e3ff8-2245-43ca-ba77-2af9af02fa07\",\"type\":\"log\",\"data\":{\"message\": \"Signal message received from http endpoint.\"}}",
+      "statusCode": 200
+    }
+  }
+]
+----
+
+The `initializerJson.json` file defines a mock http request to the path `/api/signal` with the query string parameter `code=10969`. When this request is received by the Mock Server, it will respond with a JSON body containing an `id`, `type`, and `data`. The response has a status code of 200, indicating a successful response.
```
Comment (Naros) [inline _posts/2023-07-10-custom-http-signaling-notification.adoc]: I might suggest rewording this into the paragraph before the code segment.  This will allow the definition list syntax I suggested to easily follow directly after the code block.

File: _posts/2023-07-10-custom-http-signaling-notification.adoc [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+curl -i -X POST -H "Accept:application/json" \
+    -H  "Content-Type:application/json" http://localhost:8083/connectors/ \
+    -d @register-postgres.json
+----
+
+Once the connector is successfully registered, you can review the connector logs to observe the signal events. The logs provide insights into the processing and progress of the connector, including any signal-related information. You will encounter log messages similar to the following:
+
+[source,shell]
+----
+http-signaling-notification-connect-1     | 2023-07-07 15:09:48,205 INFO   Postgres|dbserver1|postgres-connector-task  Recorded signal event 'SignalRecord{id='924e3ff8-2245-43ca-ba77-2af9af02fa07', type='log', data='{"message":"Signal message received from http endpoint."}', additionalData={}}'    [io.debezium.examples.signal.HttpSignalChannel]
```
Comment (Naros) [inline _posts/2023-07-10-custom-http-signaling-notification.adoc]:
```suggestion
Recorded signal event 'SignalRecord{id='924e3ff8-2245-43ca-ba77-2af9af02fa07', type='log', data='{"message":"Signal message received from http endpoint."}', additionalData={}}'    [io.debezium.examples.signal.HttpSignalChannel]
```
The prefix bit is a bit verbose and doesn't really provide much value from the blog's intent, so I'd suggest removing it and starting with the relevant log text.

File: _posts/2023-07-10-custom-http-signaling-notification.adoc [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+----
+
+Additionally, you might notice log messages related to notification events being sent to the Postbin. For example:
+
+[source,shell]
+----
+http-signaling-notification-connect-1     | 2023-07-07 15:09:47,737 INFO   Postgres|dbserver1|snapshot  [HTTP NOTIFICATION SERVICE] Sending notification to http channel   [io.debezium.examples.notification.HttpNotificationChannel]
+http-signaling-notification-connect-1     | 2023-07-07 15:09:48,347 INFO   Postgres|dbserver1|snapshot  Bin created: {"binId":"1688742588469-1816775151528","now":1688742588470,"expires":1688744388470}   [io.debezium.examples.notification.HttpNotificationChannel]
```
Comment (Naros) [inline _posts/2023-07-10-custom-http-signaling-notification.adoc]:
```suggestion
[HTTP NOTIFICATION SERVICE] Sending notification to http channel   [io.debezium.examples.notification.HttpNotificationChannel]
Bin created: {"binId":"1688742588469-1816775151528","now":1688742588470,"expires":1688744388470}   [io.debezium.examples.notification.HttpNotificationChannel]
```
Same as above just cut these back for a bit of brevity and clarity.

File: _posts/2023-07-10-custom-http-signaling-notification.adoc [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+
+== Conclusion
+
+In this tutorial, we explored how to create custom signal and notification channels for Debezium connectors. We created a custom signal channel that receive a signal event from an HTTP endpoint. We also created a custom notification channel that sends a notification event to an HTTP endpoint.
+
+Debezium's comprehensive signaling and notification system offers seamless integration with third-party solutions, allowing users to stay informed about the state and progress of Debezium connectors. The system's extensibility empowers users to customize both the signals and notification channels to fit their customized needs.
+
+Stay tuned for Part 3 of this series where we will explore about JMX signaling and notifications. In the meantime, you can check out the Debezium documentation for more information about signal and notification channels.
+
+If you have any questions or feedback, please feel free to reach out to us on the Debezium mailing list or the #community-general channel on the Zulip chat. We would love to hear from you!
```
Comment (Naros) [inline _posts/2023-07-10-custom-http-signaling-notification.adoc]: Perhaps embed the links to the mailing list and the Zulip channel for quick accessibility.

**Review by Naros — COMMENTED** [review]
Hi @ani-sha and @mfvitale, overall this was a fabulous read, great work.  I've left a few inline comments, thanks!

**Comment by Naros** [comment]
I'll publish this today (Friday) and tweet about this on Monday. Great work, you two! :pray:
